core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -48,21 +59,75 @@ protected RetryingClient(Client<I, O> delegate, RetryStrategy<I, O> retryStrateg
         this.defaultMaxAttempts = defaultMaxAttempts;
     }
 
-    /**
-     * Creates a new {@link Backoff} instance.
-     * If the created instance does not have {@link AttemptLimitingBackoff}, it will be wrapped with
-     * the instance of {@link AttemptLimitingBackoff}.
-     * @return the {@link Backoff} which is wrapped by {@link AttemptLimitingBackoff} if it doesn't have
-     */
-    protected Backoff newBackoff() {
+    @Override
+    public O execute(ClientRequestContext ctx, I req) throws Exception {
+        setDeadlineOfThisRequest(ctx);
+        return doExecute(ctx, req, newBackoff());
+    }
+
+    private static void setDeadlineOfThisRequest(ClientRequestContext ctx) {
+        final Attribute<Long> deadlineNanosAttr = ctx.attr(RESPONSE_TIMEOUT_DEADLINE_NANOS);
+        if (ctx.responseTimeoutMillis() <= 0) {
+            deadlineNanosAttr.set(-1L);
+        } else {
+            deadlineNanosAttr.set(
+                    System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(ctx.responseTimeoutMillis()));
+        }
+    }
+
+    @VisibleForTesting
+    Backoff newBackoff() {
         Backoff backoff = backoffSupplier.get();
         if (!backoff.as(AttemptLimitingBackoff.class).isPresent()) {
             backoff = backoff.withMaxAttempts(defaultMaxAttempts);
         }
         return backoff;
     }
 
+    /**
+     * Invoked by {@link #execute(ClientRequestContext, Request)}
+     * after the deadline for response timeout is set.
+     */
+    protected abstract O doExecute(ClientRequestContext ctx, I req, Backoff backoff) throws Exception;
+
     protected RetryStrategy<I, O> retryStrategy() {
         return retryStrategy;
     }
+
+    /**
+     * Resets the {@link ClientRequestContext#responseTimeoutMillis()}.
+     * @throws ResponseTimeoutException if the remaining response timeout is equal to or less than 0
+     */
+    protected final void resetResponseTimeout(ClientRequestContext ctx) {
+        final long responseTimeoutMillis = responseTimeoutMillis(ctx);
+        if (responseTimeoutMillis < 0) { // response timeout is disabled.
+            return;
+        }
+
+        ctx.setResponseTimeoutMillis(responseTimeoutMillis);
+    }
+
+    /**
+     * Returns the next delay which retry will be made after. The delay is the smaller value of {@code nextDelay}
+     * and {@code responseTimeoutMillis}. If response timeout is disabled, just returns {@code nextDelay}.
+     * @throws ResponseTimeoutException if the remaining response timeout is equal to or less than 0
+     */
+    protected final long getNextDelay(long nextDelay, ClientRequestContext ctx) {
```

Comment:
nit: how about changing `nextDelay` to `maxDelay` or something that means a limitation?